### PR TITLE
Issue #SH-636 feat: added consolidated view of progress

### PR DESCRIPTION
**Pull request by swetha-aggidevara**

### CODE REVIEW

File: src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts [hunk 1]
```diff
@@ -1,28 +1,102 @@
...
 @Component({
   selector: 'app-course-dashboard',
   templateUrl: './course-dashboard.component.html',
   styleUrls: ['./course-dashboard.component.scss']
 })
 export class CourseDashboardComponent implements OnInit, OnDestroy {
-  batchList = [];
-  dashBoardData: IDashboard;
+
+  public dashBoardSmallCards: IDashboard = {
```
Comment (sudip-mukherjee) [inline src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:26]: change the variable name

File: src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts [hunk 2]
```diff
@@ -1,28 +1,102 @@
...
+      title: dashBoardTitles.batches,
+      count: 0
+    },
+    totalEnrollment: {
+      title: dashBoardTitles.enrollments,
+      count: 0
+    }
+  };
+
+  public dashBoardLargeCards: IDashBoardLarge = {
```
Comment (sudip-mukherjee) [inline src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:37]: change the variable name  here as well

File: src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts [hunk 3]
```diff
@@ -1,28 +1,102 @@
...
+    totalCompleted: {
+      title: dashBoardTitles.completions,
+      count: 0
+    }
+  };
+
   public unsubscribe$ = new Subject<void>();
+  courseId: string;
+  queryParams;
```
Comment (sudip-mukherjee) [inline src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:46]: add type to queryParams

File: src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts [hunk 4]
```diff
@@ -1,28 +1,102 @@
...
+      context: {
+        env: _.get(this.activatedRoute.snapshot, 'data.telemetry.env')
+      },
+      edata: {
+        type: _.get(this.activatedRoute.snapshot, 'data.telemetry.type'),
+        pageid: _.get(this.activatedRoute.snapshot, 'data.telemetry.pageid'),
+        uri: this.router.url,
+        duration: this.navigationhelperService.getPageLoadTime()
+      }
+    };
```
Comment (sudip-mukherjee) [inline src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:75]: share the telemetry event on instrumentation sheet and ask Vinu sir to review

File: src/app/client/src/app/modules/dashboard/components/dashboard-sidebar/dashboard-sidebar.component.html [hunk 5]
```diff
@@ -2,9 +2,10 @@
   <div class="three wide column">
     <div class="sidebar py-24">
       <div class="dashboard-title m-0 mb-16">Dashboard</div>
-      <a class="active" href="#progress">Progress</a>
-      <a href="#batches">Batches</a>
-      <a href="#certificates">Certificates</a>
+      <a [routerLink]="['course-stats']" routerLinkActive="active" appTelemetryInteract [telemetryInteractEdata]="courseStatsEdata" >Progress</a>
+      <a [routerLink]="['progress']" [class.active]="router.url.includes('/dashboard/progress')" appTelemetryInteract [telemetryInteractEdata]="courseBatchesEdata">Batches</a>
+      <a [routerLink]="['certificates']" [class.active]="router.url.includes('/dashboard/certificates')" appTelemetryInteract [telemetryInteractEdata]="courseCertificatesEdata">Certificates</a>
+
```
Comment (sudip-mukherjee) [inline src/app/client/src/app/modules/dashboard/components/dashboard-sidebar/dashboard-sidebar.component.html:8]: remove the hardcoded labels  Progress, Batches and Certificates

File: src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts [hunk 6]
```diff
@@ -1,28 +1,102 @@
...
 import { CourseProgressService } from '../../services';
+import { ActivatedRoute, Router } from '@angular/router';
+import {  } from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
+
+enum dashBoardTitles {
+  batches = 'TOTAL BATCHES',
+  enrollments = 'TOTAL ENROLLMENTS',
+  completions = 'TOTAL COMPLETIONS',
+}
```
Comment (sudip-mukherjee) [inline src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:18]: If we set the labels like this, will it support language change?

File: src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts [hunk 7]
```diff
@@ -1,28 +1,102 @@
...
 
   getBatchList() {
+    const searchParams = {
+      courseId: this.courseId,
+      status: ['0', '1', '2'],
+      createdBy: this.userService.userid
+    };
+    this.courseProgressService.getBatches(searchParams).pipe(takeUntil(this.unsubscribe$)).subscribe(data => {
+      this.getEnrollmentAndCompletedCount(_.get(data, 'result.response'));
+    }, err => {
```
Comment (sudip-mukherjee) [inline src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:86]: what will happen if the getBatches() api fails ? no error message will be shown, right ?

File: src/app/client/src/app/modules/dashboard/components/dashboard-sidebar/dashboard-sidebar.component.html (hunk 5 above)
Comment (sudip-mukherjee) [inline src/app/client/src/app/modules/dashboard/components/dashboard-sidebar/dashboard-sidebar.component.html:8]: instrument the events and ask vinu to review

File: src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts (hunk 7 above)
Comment (sudip-mukherjee) [inline src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:86]: test cases for this ?

File: src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts [hunk 8]
```diff
@@ -1,28 +1,102 @@
...
-  getEnrollmentAndCompletedCount() {
+  getEnrollmentAndCompletedCount(batchList) {
+    const batches = _.get(batchList, 'content');
+    if (batches) {
+      this.dashBoardSmallCards.totalBatches.count = _.get(batchList, 'count');
+      _.forEach(batches, batch => {
+        this.dashBoardLargeCards.totalCompleted.count += _.get(batch, 'completedCount');
+        this.dashBoardSmallCards.totalEnrollment.count += _.get(batch, 'participantCount');
+      });
+    }
```
Comment (sudip-mukherjee) [inline src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:99]: test cases for this method as well. and add DOC comments for each and every method. will be easier to understand.

File: src/app/client/src/app/modules/dashboard/interfaces/dashboardData.ts [hunk 9]
```diff
@@ -59,11 +59,14 @@ export interface IBatchListData {
...
-    count: number;
+    count: Number;
 }
 
 export interface IDashboard {
     totalBatches: IDashboardItem;
     totalEnrollment: IDashboardItem;
+}
+
+export interface IDashBoardLarge {
```
Comment (sudip-mukherjee) [inline src/app/client/src/app/modules/dashboard/interfaces/dashboardData.ts:70]: change the interface name. IDashBoardLarge

File: src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts [hunk 10]
```diff
@@ -1,28 +1,118 @@
...
+    this.state = history.state;
+    this.initializeFields();
+    this.courseId = _.get(this.activatedRoute.snapshot.params, 'courseId') || _.get(this.state, 'id');
+    this.queryParams = this.activatedRoute.snapshot.queryParams;
+    this.getBatchList();
+    this.setImpressionEvent();
   }
 
+  // To initialize the dashboard data
+  initializeFields() {
```
Comment (sudip-mukherjee) [inline src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:50]: add language change subscription.

Comment (sudip-mukherjee) [inline src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:50]: add language change subscription.

File: src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts [hunk 11]
```diff
@@ -1,28 +1,120 @@
...
+    this.getBatchList();
+    this.setImpressionEvent();
+  }
+
+  // To initialize the dashboard data
+  initializeFields() {
+    this.resourceService.languageSelected$.pipe(takeUntil(this.unsubscribe$))
+    .subscribe(item => {
+    this.dashBoardItems = {
+      totalBatches: {
```
Comment (vinukumar-vs) [inline src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:53]: This should be IDashboard instance. The title is already part of each object, then why do you want 'totalBatches' as another key in dashboardItems?
Yout HTML should be for loop of this.dashBoardItems & show the cards in the same object with type of card.

File: src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts [hunk 12]
```diff
@@ -1,28 +1,120 @@
...
+      this.getEnrollmentAndCompletedCount(_.get(data, 'result.response'));
+    }, err => {
+      this.toasterService.error(this.resourceService.messages.emsg.m0005);
+    });
+
   }
 
-  getEnrollmentAndCompletedCount() {
+  // To get the user enrolled/completed  count for the Course
+  getEnrollmentAndCompletedCount(batchList) {
```
Comment (vinukumar-vs) [inline src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:109]: rename getEnrollmentAndCompletedCount(batchList)  -> getDashboardData(batchList)

File: src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts [hunk 13]
```diff
@@ -1,28 +1,120 @@
...
+    const batches = _.get(batchList, 'content');
+    if (batches) {
+      this.dashBoardItems.totalBatches.count = _.get(batchList, 'count');
```
Comment (vinukumar-vs) [inline src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:112]: Here you create IDashboard instance for each count & push it to this.dashboardItems.

File: src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts (hunk 11 above)
Comment (swetha-aggidevara) [reply src/app/client/src/app/modules/dashboard/components/course-dashboard/course-dashboard.component.ts:53]: This is to assign the string data

**Pull request merged**